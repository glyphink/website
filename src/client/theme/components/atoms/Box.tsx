/**
 * @file Components - Box
 * @module website/theme/components/atoms/Box
 */

import type { Props } from '@glyphink/website/theme'
import {
  forwardRef,
  type ForwardedRef,
  type ForwardRefExoticComponent,
  type JSX
} from 'react'
import type BoxProps from './Box.props.mts'

/**
 * A `<div>` element.
 *
 * - https://developer.mozilla.org/docs/Web/API/HTMLDivElement
 *
 * @see {@linkcode BoxProps}
 * @see {@linkcode ForwardRefExoticComponent}
 *
 * @const {ForwardRefExoticComponent<BoxProps>} Box
 */
const Box: ForwardRefExoticComponent<BoxProps> = forwardRef<
  HTMLDivElement,
  BoxProps
>(
  /**
   * @param {Props<BoxProps>} props
   *  Component props
   * @param {ForwardedRef<HTMLDivElement>} ref
   *  Forwarded ref
   * @return {JSX.Element}
   *  JSX `<div>` element
   */
  (props: Props<BoxProps>, ref: ForwardedRef<HTMLDivElement>): JSX.Element => {
    return <div {...props} ref={ref} />
  }
)

Box.displayName = 'Box'

export default Box
