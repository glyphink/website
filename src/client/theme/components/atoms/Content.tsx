/**
 * @file Components - Content
 * @module website/theme/components/atoms/Content
 */

import type { Props } from '@glyphink/website/theme'
import cx from 'classnames'
import { ok } from 'devlop'
import {
  forwardRef,
  type ForwardedRef,
  type ForwardRefExoticComponent,
  type JSX
} from 'react'
import Box from './Box'
import classes from './Content.module.scss'
import type ContentProps from './Content.props.mts'

/**
 * Page content.
 *
 * @see {@linkcode ContentProps}
 * @see {@linkcode ForwardRefExoticComponent}
 *
 * @const {ForwardRefExoticComponent<ContentProps>} Content
 */
const Content: ForwardRefExoticComponent<ContentProps> = forwardRef<
  HTMLDivElement,
  ContentProps
>(
  /**
   * @param {Props<ContentProps>} props
   *  Component props
   * @param {ForwardedRef<HTMLDivElement>} ref
   *  Forwarded ref
   * @return {JSX.Element}
   *  `Box` element
   */
  (
    props: Props<ContentProps>,
    ref: ForwardedRef<HTMLDivElement>
  ): JSX.Element => {
    ok(classes.content, 'expected `classes.content`')

    return (
      <Box
        {...props}
        className={cx(classes.content, props.className)}
        id='content'
        ref={ref}
      />
    )
  }
)

Content.displayName = 'Content'

export default Content
