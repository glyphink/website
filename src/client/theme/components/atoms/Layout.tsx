/**
 * @file Components - Layout
 * @module website/theme/components/atoms/Layout
 */

import type { Props } from '@glyphink/website/theme'
import cx from 'classnames'
import { ok } from 'devlop'
import type { FC, JSX } from 'react'
import Box from './Box.tsx'
import Content from './Content.tsx'
import classes from './Layout.module.scss'
import type LayoutProps from './Layout.props.mts'

/**
 * Page layout.
 *
 * @see {@linkcode LayoutProps}
 *
 * @type {FC<LayoutProps>}
 *
 * @param {Props<LayoutProps>} props
 *  Component props
 * @return {JSX.Element}
 *  `Box` element
 */
const Layout: FC<LayoutProps> = (props: Props<LayoutProps>): JSX.Element => {
  ok(classes.layout, 'expected `classes.layout`')

  return (
    <Box {...props} className={cx(classes.layout, props.className)}>
      <Content>
        {props.children}
      </Content>
    </Box>
  )
}

Layout.displayName = 'Layout'

export default Layout
