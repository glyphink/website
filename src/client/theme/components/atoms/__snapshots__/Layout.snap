// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:theme/components/atoms/Layout > should render <div> 1`] = `
<div
  class="layout"
>
  <div
    class="content"
    id="content"
  />
</div>
`;
